### DISCUSSION TITLE
for the training process, loss range detection

### DISCUSSION THREAD

**Discussion by DDDPG** [speechbrain/speechbrain#2840] [Q&A]
When training the model, the system always said like:
```
__main__ - infinite loss or empty loss! it happened 1 times so far - skipping this batch     
__main__ - infinite loss or empty loss! it happened 2 times so far - skipping this batch     
__main__ - infinite loss or empty loss! it happened 3 times so far - skipping this batch     
__main__ - infinite loss or empty loss! it happened 4 times so far - skipping this batch     
...
```
But when testing the compute_objectives() in individually like:
```
test_loader = sb.dataio.dataloader.make_dataloader(
        test_data, **separator.hparams.dataloader_opts
    )
    with tqdm(test_loader, dynamic_ncols=True) as t:
        for i, batch in enumerate(t):
            mix_sig = batch.mix_sig[0]
            human_sig = batch.human_sig[0]
            robot_sig = batch.robot_sig[0]
            robot_length = torch.tensor([robot_sig.shape[1]] * robot_sig.shape[0])
            predictions, targets = separator.compute_forward(
                mix_sig, robot_sig, robot_length, human_sig, sb.Stage.TEST
            )
            loss = separator.compute_objectives(predictions, targets, sb.Stage.TEST)
            print(loss, "loss{}".format(i))
```

It could return normal loss value, so how does the Brain class judge if the loss is empty or infinite?

**Comment by DDDPG** [comment]
further check update:

the if-else case for judging the loss is in the fit_batch() for the recipe: `loss.nelement() > 0 and loss < self.hparams.loss_upper_lim`, but something still wired:
* in the test case for compute_objectives(), the loss could be valid
* but when running the real fit() process and print the `loss.nelement()` and `loss`, it always `tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)`

Could any step lead to this situation?

the recipe is `WHAMandWHAMR/separation/train.py`

thanks a lot!

**Reply by DDDPG** [reply]
Additional explanation: I modified the `loss` used in yaml to a locally written function named `models.RoTFnet.stftloss.getcombinedloss`. It can calculate the loss normally when testing `compute_objectives()` separately.
